[WIP] Add new varlong data type and test varint

deathcap opened the pull request:
This could be useful for https://github.com/PrismarineJS/minecraft-data/issues/119

From http://wiki.vg/Protocol#Data_types :

| Name | Size | Encodes | Notes |
| --- | --- | --- | --- |
| VarInt | ≥ 1 ≤ 5 | An integer between -2147483648 and 2147483647 | [Protocol Buffer Varint](http://developers.google.com/protocol-buffers/docs/encoding#varints), encoding a two's complement signed 32-bit integer |
| VarLong | ≥ 1  ≤ 10 | An integer between -9223372036854775808 and 9223372036854775807 | [Protocol Buffer Varint](http://developers.google.com/protocol-buffers/docs/encoding#varints), encoding a two's complement signed 64-bit integer |

The current `varint` in ProtoDef allows >5-byte encodings (>32-bit), but overflows JavaScript's 32-bit bitwise arithmetic operations and 53-bit number type

rom1504 commented:
Still [WIP] ?

Anyway, I'm going to npm publish ProtoDef with my numerical type change, that will be for next release.

deathcap commented:
Very [WIP], currently broken because of JavaScript's lack of 64-bit data types (float is 53-bit) but worse, bitwise operators coerce to 32-bit so I can't even use the 53-bits, as written. No more broken than before where varint would read but overflow >32-bit, but I think this PR should be converted to use something like https://www.npmjs.com/package/node-int64 before pulling. Also simultaneously updating world_border in minecraft-data https://github.com/PrismarineJS/minecraft-data/issues/119 to varlong speed, since this PR does restrict varint to 32-bit, which could cause world_border to fail to parse unless it is changed to varlong

deathcap commented:
Note: rebases cleanly on master (with the numerical type change) but then tests fail, not yet investigated

rom1504 commented:
For long (now i64) we currently use an array of 2 numbers. It works but it's obviously not quite convenient.

deathcap commented:
An array of two 32-bit numbers is about the best one can do with JavaScript until a true 64-bit type is added, node-int64 basically just wraps it the 2 x 32-bit array in a nice class (there are modules to wrap C's int64_t, but require native code). To do: update this PR for varlong to return 64-bit as [high, low].

deathcap commented:
Found this useful module: https://github.com/pocka/bitwise64 - gets around the limitation of JavaScript bitwise &|~ coercing to 32-bit, provides its own bitwise64.and/or/not functions. Still works within the confines of 53-bit JavaScript arithmetic, but makes it easier to port the varint parsing code, as it makes heavy use of bitwise operations. Fixed and added more tests, now is able to read, sizeof, and write up to 0x7fffffffffffff (36028797018963967).

Negatives are problematic however. If I'm understanding the varlong format correctly, then -1 should be encoded as 10 bytes. -1 encoded as varint is 5-bytes (ff ff ff ff 0f), but ff ff ff ff 0f decoded as varlong is 4294967295 (0xffffffff) — I have this working, but encoding -1 as 'ff ff ff ff ff ff ff ff ff 0f' would require more than 53-bit precision at least with the current implementation technique. 

Pushed what I have so far (also idk why the CircleCI tests are failing `Error: Cannot find module './datatypes/utils'`, they work here locally), but I think what we really need to implement varlong correctly is 64-bit [bitwise operations](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Bitwise_Operators) (and, or, not, left-shift, right-shift-zero-fill) on pairs of 32-bit integers. Not aware of such a module. All I've found is:

* [bitwise64](https://www.npmjs.com/package/bitwise64) - only works up to 53-bit (floating point mantissa)
* [int53](https://www.npmjs.com/package/int53) - reads/writes 64-bit buffers, but as 53-bit
* [node-int64](https://www.npmjs.com/package/node-int64) - converts to 2x32, but no arithmetic
* https://www.npmjs.com/browse/keyword/int64
* https://www.npmjs.com/browse/keyword/bitwise

roblabla commented:
Same response as in #52, could we please have a general varnum instead of having more and more datatypes ?

deathcap commented:
Could be possible to add a varnum with a maximum bytes type argument, but note that anything max bytes > 5 (bigger than a varint) would need the different (varlong) code path, unless we want to remove the 32-bitwise operations in the varint code path as well but those are faster than 53-bit/64-bit arithmetic.

varlong (10-byte) also needs a different value type (array of [32-bit, 32-bit] to cover the full 64-bit precision, versus a simple numeric type for varint), "shorter" varlongs may or may not need this different type representation (since can store up to 53-bit numbers in JavaScript), but they will need the 53- or 64-bit arithmetic since JavaScript bitwise operations coerce to 32-bit. The other difference is max bytes is also the number of bytes used for -1 (etc.).

Enumerating the possibilities:

type(ish) | max bytes | codec | JS value type
--- | --- | --- | --- | ---
varint | 1 | 32-bit bitwise | number (32-bit)
varint | 2 | 32-bit bitwise | number (32-bit)
varint | 3 | 32-bit bitwise | number (32-bit)
varint | 4 | 32-bit bitwise | number (32-bit)
varint | 5 | 32-bit bitwise | number (32-bit)
varlong | 6 | 53/64-bit arithmetic | floating number (>32-bit)
varlong | 7 | 53/64-bit arithmetic | floating number (>32-bit)
varlong | 8 | 53/64-bit arithmetic | floating number (>32-bit)
varlong | 9 | 53/64-bit arithmetic | pair of \[32-bit, 32-bit\] ([node-int64](https://www.npmjs.com/package/node-int64) / ProtoDef readLong)
varlong | 10 | 53/64-bit arithmetic | pair of \[32-bit, 32-bit\] ([node-int64](https://www.npmjs.com/package/node-int64) / ProtoDef readLong)

It is an interesting idea, but to support MC, I think all we really need are varint (5) and varlong (10) http://wiki.vg/Protocol#Data_types. Going to be afk for a bit so I won't able to work on this soon, but maybe this refactoring could be done after the 64-bit and negative varlong tests pass (still need a full 64-bit module, current very incomplete progress (unpublished): [bitwise2x32](https://github.com/deathcap/bitwise2x32))

roblabla commented:
Right, you have a valid point concerning the [x, y] array for varlongs (had completely forgotten about that stupid limitation). So I guess varlongs are OK.

The thing is, a part of me doesn't want to add more datatypes to ProtoDef's core than already exists. IMO, varint is already borderline "too much". Having a specialized varlong that's unlikely to be used anywhere other than minecraft bothers me. And the fact that it returns an array of [int32, int32], making it not fit into JS easily, bothers me even more.

I think it might be a good idea to separate those into small modules (protodef-varlong, protodef-varshort ?)  that exposes the datatypes for those that want them. I'd like protodef's core to be kept to datatypes that are useful for describing common binary protocols (think, HTTP2, SSH, etc...) without having special code to handle every single obscure unknown protocol quirks.

ProtoDef is something that's intended to be used in a variety of projects, not just NMP.

filfat commented:
bump :)

rom1504 commented:
seems like we haven't needed this
if someone need this one day, he can reopen
